packages/dashboard/src/components/iot-dashboard/iot-context-menu/iot-dashboard-context-menu.tsx

```diff
@@ -0,0 +1,116 @@
+  @Prop() deleteWidgets: () => void;
+  @Prop() bringToFront: () => void;
+  @Prop() sendToBack: () => void;
+
+  /** Widget selections to be used to determine enabled actions. */
+  @Prop() hasSelectedWidgets: boolean;
+  @Prop() hasCopiedWidgets: boolean;
+
+  /** State to keep track of the context menu popover visibility and position. */
+  @State() contextMenuOpen = false;
+  @State() contextMenuPosition: Position | undefined;
+
+  private contextMenuTarget: HTMLElement | null;
+
+  componentWillLoad() {
+    this.contextMenuTarget = document.querySelector('iot-dashboard-context-menu')?.parentElement ?? null;
+
+    if (this.contextMenuTarget) {
+      this.contextMenuTarget.addEventListener('contextmenu', this.onContextMenu.bind(this));
+      this.contextMenuTarget.addEventListener('mousedown', this.onMouseDown.bind(this));
+    }
+  }
+
+  disconnectedCallback() {
+    if (this.contextMenuTarget) {
+      this.contextMenuTarget.removeEventListener('contextmenu', this.onContextMenu.bind(this));
+      this.contextMenuTarget.removeEventListener('mousedown', this.onMouseDown.bind(this));
+    }
+  }
+
+  onContextMenu(event: MouseEvent) {
+    const { x, y } = getDashboardPosition(event);
+    this.contextMenuOpen = !this.contextMenuOpen;
+    this.contextMenuPosition = { x, y };
+    event.preventDefault();
+    return;
+  }
+
+  onMouseDown(event: MouseEvent) {
+    const isContextMenuGesture = event.button === MouseClick.Right;
+
+    if (!isContextMenuGesture && this.contextMenuOpen) {
+      this.hideContextMenu();
+    }
+  }
+
+  hideContextMenu() {
+    this.contextMenuOpen = false;
+    this.contextMenuPosition = undefined;
+  }
+
+  getContextMenuConfiguration = (): ContextMenuConfiguration => {
+    const withClose = (action: (event: MouseEvent) => void) => (event: MouseEvent) => {
+      action(event);
+      this.hideContextMenu();
+    };
+
+    const copyAction = withClose(() => this.copyWidgets());
+    const pasteAction = withClose(() => this.pasteWidgets(this.contextMenuPosition));
+    const deleteAction = withClose(() => this.deleteWidgets());
+    const bringToFrontAction = withClose(() => this.bringToFront());
+    const sendToBackAction = withClose(() => this.sendToBack());
+
+    return createContextMenuOptions({
+      messages: this.messageOverrides.contextMenu,
+      actions: {
+        copyAction,
+        pasteAction,
+        deleteAction,
+        bringToFrontAction,
+        sendToBackAction,
+      },
+      state: {
+        hasCopiedWidgets: this.hasCopiedWidgets,
+        hasSelectedWidgets: this.hasSelectedWidgets,
+      },
+    });
+  };
+
+  render() {
+    if (this.contextMenuOpen && this.contextMenuPosition) {
+      return (
+        <iot-context-menu {...this.contextMenuPosition}>
+          {this.getContextMenuConfiguration().map(({ id, options }) => (
+            <iot-context-menu-section key={id}>
+              {options.map(({ id, text, hotkey, action, disabled }) => (
```

Comment:
has shadow variable `id` (should be caught by eslint, but that's another story). should use a variable alias i.e. `id: optionId`

Reply:
yup good catch. not sure how the linter didn't catch that